CppImport/src/macro/ClangHelper.cpp

```diff
@@ -0,0 +1,126 @@
+/***********************************************************************************************************************
+ **
+ ** Copyright (c) 2011, 2015 ETH Zurich
+ ** All rights reserved.
+ **
+ ** Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+ ** following conditions are met:
+ **
+ **    * Redistributions of source code must retain the above copyright notice, this list of conditions and the
+ **      following disclaimer.
+ **    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+ **      following disclaimer in the documentation and/or other materials provided with the distribution.
+ **    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+ **      derived from this software without specific prior written permission.
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "ClangHelper.h"
+
+namespace CppImport {
+
+void ClangHelper::setSourceManager(const clang::SourceManager* sourceManager)
+{
+	sourceManager_ = sourceManager;
+}
+
+void ClangHelper::setPreprocessor(const clang::Preprocessor* preprocessor)
+{
+	preprocessor_ = preprocessor;
+}
+
+QString ClangHelper::spelling(clang::SourceLocation loc)
+{
+	return spelling(loc, loc);
+}
+
+QString ClangHelper::spelling(clang::SourceRange range)
+{
+	return spelling(range.getBegin(), range.getEnd());
+}
+
+QString ClangHelper::spelling(clang::SourceLocation start, clang::SourceLocation end)
+{
+	clang::SourceLocation b = sourceManager_->getSpellingLoc(start);
+	clang::SourceLocation e = clang::Lexer::getLocForEndOfToken(sourceManager_->getSpellingLoc(end), 0, *sourceManager_,
+																					preprocessor_->getLangOpts());
+
+	auto length = sourceManager_->getCharacterData(e) - sourceManager_->getCharacterData(b);
+	if (length > 1000000) return "ERROR_IN_GET_SPELLING";
```

Comment:
Errr. Returning an error string? Really? Is this error expected every now and then? If not, the use `Q_ASSERT`. If yes, then it's not really an error and we should probably handle it differently.